synapse/handlers/message.py

```diff
@@ -1531,12 +1531,18 @@ async def cache_joined_hosts_for_events(
         external federation senders don't have to recalculate it themselves.
         """
 
-        for event, _ in events_and_context:
-            if not self._external_cache.is_enabled():
-                return
+        if not self._external_cache.is_enabled():
+            return
+
+        # If external cache is enabled we should always have this.
+        assert self._external_cache_joined_hosts_updates is not None
 
-            # If external cache is enabled we should always have this.
-            assert self._external_cache_joined_hosts_updates is not None
+        for event, event_context in events_and_context:
+            if event_context.partial_state:
+                # We can't precalculate joined hosts for a partial-state event,
+                # (at least not without blocking until full state).
+                # So skip this calculation entirely.
```

Comment:
I think this is slightly confusing wording. You're right that we can't calculate the hosts from the state, but we *do* know the hosts we'll send the event to (albeit we'd have to query the DB) and so could add that to the external cache.

There's probably not much point doing so though.

Reply:
How does the following sound?

```diff
-                 # We can't precalculate joined hosts for a partial-state event,
-                 # (at least not without blocking until full state).
-                 # So skip this calculation entirely.
+                # The federation senders don't use the external cache when sending
+                # events in partial-state rooms, so there's not much point in
+                # precalculating the hosts to send to.
+                #
+                # Note that the precalculation logic below will block for partial-state
+                # events and so can't be used for such events.
```

Reply:
wfm!

Reply:
Reworded it again, to
```diff
-                # We can't precalculate joined hosts for a partial-state event,
-                # (at least not without blocking until full state).
-                # So skip this calculation entirely.
+                # To populate the cache for a partial-state event, we either have to
+                # block until full state, which the code below does, or change the
+                # meaning of cache values to be the list of hosts to which we plan to
+                # send events and calculate that instead.
+                #
+                # The federation senders don't use the external cache when sending
+                # events in partial-state rooms anyway, so let's not bother populating
+                # the cache.
```